lib/peer/router.ex

```diff
@@ -0,0 +1,167 @@
+defmodule ElixiumNode.Router do
+  use Pico.Client.Router
+  require Logger
+  alias Elixium.Node.LedgerManager
+  alias Elixium.Store.Ledger
+  alias Elixium.Pool.Orphan
+  alias Elixium.Block
+  alias Elixium.Transaction
+  alias Elixium.Validator
+  alias Elixium.Store.Oracle
+
+  message "BLOCK", block do
+    block = Block.sanitize(block)
+
+    case LedgerManager.handle_new_block(block) do
+      :ok ->
+        # We've received a valid block. We need to gossip this block to all the
+        # nodes we know of.
+        Logger.info("Received valid block #{block.hash} at index #{:binary.decode_unsigned(block.index)}.")
+        Pico.broadcast("BLOCK", block)
+
+        known_transactions = SharedState.get(:transactions) || [] -- block.transactions
+        SharedState.set(:transactions, known_transactions)
+
+      :gossip ->
+        # For one reason or another, we want to gossip this block. (Perhaps this is a fork block)
+        Pico.broadcast("BLOCK", block)
+
+      {:missing_blocks, fork_chain} ->
+        # We've discovered a fork, but we can't rebuild the fork chain without
+        # some blocks. Let's request them from our peer.
+        Pico.message(conn, "BLOCK_QUERY_REQUEST", %{index: :binary.decode_unsigned(hd(fork_chain).index) - 1})
+
+      :invalid ->
+        Logger.info("Recieved invalid block at index #{:binary.decode_unsigned(block.index)}.")
+
+      :ignore -> :ignore
+    end
+  end
+
+  message "BLOCK_QUERY_REQUEST", %{index: index} do
+    Pico.message(conn, "BLOCK_QUERY_RESPONSE", Ledger.block_at_height(index))
+  end
+
+  message "BLOCK_QUERY_RESPONSE", response do
+    orphans_ahead =
+      Ledger.last_block().index
+      |> :binary.decode_unsigned()
+      |> Kernel.+(1)
+      |> Orphan.blocks_at_height()
+      |> length()
+
+    if orphans_ahead > 0 do
```

Comment:
We should start highlighting this stuff to be removed, if we dont use it within a certain time frame

Reply:
This is a TODO, left it here so we don't forget to implement it